ibm_is_security_group_target should wait if target loadbalancer is in update_pending state

<!---
Please note the following potential times when an issue might be in Terraform core:

* [Configuration Language](https://www.terraform.io/docs/configuration/index.html) or resource ordering issues
* [State](https://www.terraform.io/docs/state/index.html) and [State Backend](https://www.terraform.io/docs/backends/index.html) issues
* [Provisioner](https://www.terraform.io/docs/provisioners/index.html) issues
* [Registry](https://registry.terraform.io/) issues
* Spans resources across multiple providers

If you are running into one of these scenarios, we recommend opening an issue in the [Terraform core repository](https://github.com/hashicorp/terraform/) instead.
--->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Terraform CLI and Terraform IBM Provider Version

<!--- Please run `terraform -v` to show the Terraform core version and provider version(s). If you are not running the latest version of Terraform or the provider, please upgrade because your issue may have already been fixed. [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions). --->

### Affected Resource(s)

<!--- Please list the affected resources and data sources. --->

* ibm_is_security_group_target

### Terraform Configuration Files

<!--- Information about code formatting: https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code --->

Please include all Terraform configurations required to reproduce the bug. Bug reports without a functional reproduction may be closed without investigation.

Reproducible when the lb is in pending update state. 

![Image](https://github.com/user-attachments/assets/f0ef699d-6d7f-45d3-96d5-806f4adbb118)

This can for instance be reproduce by attaching 2 different SG to the same lb using ibm_is_security_group_target resource twice.

### Debug Output

<!---
Please provide a link to a GitHub Gist containing the complete debug output. Please do NOT paste the debug output in the issue; just paste a link to the Gist.

To obtain the debug output, see the [Terraform documentation on debugging](https://www.terraform.io/docs/internals/debugging.html).
--->

### Panic Output

<!--- If Terraform produced a panic, please provide a link to a GitHub Gist containing the output of the `crash.log`. --->

 Terraform apply |   [ERROR] Error while creating Security Group Target Binding The load balancer with ID 'xxxx-xxxx-xxxx' cannot be updated because its status is 'UPDATE_PENDING'.

### Expected Behavior

<!--- What should have happened? --->

ibm_is_security_group_target should wait until the lb is in state active up to a specific amount of time (eg: 5 minutes, 
should be configurable). on create and destroy.

### Actual Behavior

<!--- What actually happened? --->

error immediately raised: 
 Terraform apply |   [ERROR] Error while creating Security Group Target Binding The load balancer with ID 'xxxx-xxxx-xxxx' cannot be updated because its status is 'UPDATE_PENDING'.

### Steps to Reproduce

See above

<!--- Please list the steps required to reproduce the issue. --->

1. `terraform apply`

### Important Factoids

<!--- Are there anything atypical about your accounts that we should know? For example: Running in EC2 Classic? --->

### References

<!---
Information about referencing Github Issues: https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests

Are there any other GitHub issues (open or closed) or pull requests that should be linked here? Vendor documentation? For example:
--->

Worked around (partially) the issue through a bash script at https://github.com/terraform-ibm-modules/terraform-ibm-base-ocp-vpc/blob/fe1a2c5565a8a288b026f800a83f048ec14ea6e6/main.tf#L610 . This is a workaround, and does not always work as using distinct resources in terraform definition. The workaround is also difficult to adapt on destroy path.

--- a/ibm/service/vpc/resource_ibm_is_security_group_target.go
+++ b/ibm/service/vpc/resource_ibm_is_security_group_target.go
@@ -114,6 +114,22 @@ func resourceIBMISSecurityGroupTargetCreate(d *schema.ResourceData, meta interfa
 
 	sg, response, err := sess.CreateSecurityGroupTargetBinding(createSecurityGroupTargetBindingOptions)
 	if err != nil || sg == nil {
+
+		if strings.Contains(strings.ToLower(err.Error()), "load balancer") && ((strings.Contains(strings.ToUpper(err.Error()), "UPDATE_PENDING")) || (strings.Contains(strings.ToUpper(err.Error()), "CREATE_PENDING"))) {
+			log.Printf("[INFO] Load balancer with ID '%s' is in UPDATE_PENDING state. Waiting for it to become available before retrying...", targetID)
+
+			// Wait for the load balancer to become available
+			_, waitErr := isWaitForSGTargetLBAvailable(sess, targetID, d.Timeout(schema.TimeoutCreate))
+			if waitErr != nil {
+				return fmt.Errorf("[ERROR] Error waiting for load balancer to become available while creating Security Group Target Binding: %s", waitErr)
+			}
+
+			sg, response, err = sess.CreateSecurityGroupTargetBinding(createSecurityGroupTargetBindingOptions)
+			// Check for errors after the initial attempt or retry
+			if err != nil || sg == nil {
+				return fmt.Errorf("[ERROR] Error while creating Security Group Target Binding %s\n%s", err, response)
+			}
+		}
 		return fmt.Errorf("[ERROR] Error while creating Security Group Target Binding %s\n%s", err, response)
 	}
 	sgtarget := sg.(*vpcv1.SecurityGroupTargetReference)
@@ -367,5 +383,40 @@ func isVNISgTargetRefreshFunc(vpcClient *vpcv1.VpcV1, vniId string) resource.Sta
 			return vni, *vni.LifecycleState, fmt.Errorf("Network Interface creationg failed with status %s ", *vni.LifecycleState)
 		}
 		return vni, *vni.LifecycleState, nil
+	}
+}
+
+// New function to wait for a load balancer to become available before attaching a security group
+func isWaitForSGTargetLBAvailable(sess *vpcv1.VpcV1, lbId string, timeout time.Duration) (interface{}, error) {
+	log.Printf("Waiting for load balancer (%s) to be available before attaching security group.", lbId)
+
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{"retry", isLBProvisioning, "update_pending"},
+		Target:     []string{isLBProvisioningDone, ""},
+		Refresh:    isSGTargetLBRefreshFunc(sess, lbId),
+		Timeout:    timeout,
+		Delay:      10 * time.Second,
+		MinTimeout: 10 * time.Second,
+	}
+
+	return stateConf.WaitForState()
+}
+
+// Refresh function for checking load balancer status before security group attachment
+func isSGTargetLBRefreshFunc(sess *vpcv1.VpcV1, lbId string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		getlboptions := &vpcv1.GetLoadBalancerOptions{
+			ID: &lbId,
+		}
+		lb, response, err := sess.GetLoadBalancer(getlboptions)
+		if err != nil {
+			return nil, "", fmt.Errorf("[ERROR] Error Getting Load Balancer : %s\n%s", err, response)
+		}
+
+		if *lb.ProvisioningStatus == "active" || *lb.ProvisioningStatus == "failed" {
+			return lb, isLBProvisioningDone, nil
+		}
+
+		return lb, isLBProvisioning, nil
 	}
 }

--- a/ibm/service/vpc/resource_ibm_is_security_group_target.go
+++ b/ibm/service/vpc/resource_ibm_is_security_group_target.go
@@ -111,6 +111,9 @@ func resourceIBMISSecurityGroupTargetCreate(d *schema.ResourceData, meta interfa
 	createSecurityGroupTargetBindingOptions := &vpcv1.CreateSecurityGroupTargetBindingOptions{}
 	createSecurityGroupTargetBindingOptions.SecurityGroupID = &securityGroupID
 	createSecurityGroupTargetBindingOptions.ID = &targetID
+	isSGTargetPrefixKey := "security_group_key_" + targetID
+	conns.IbmMutexKV.Lock(isSGTargetPrefixKey)
+	defer conns.IbmMutexKV.Unlock(isSGTargetPrefixKey)
 
 	sg, response, err := sess.CreateSecurityGroupTargetBinding(createSecurityGroupTargetBindingOptions)
 	if err != nil || sg == nil {
@@ -193,6 +196,12 @@ func resourceIBMISSecurityGroupTargetRead(d *schema.ResourceData, meta interface
 	if target.ResourceType != nil && *target.ResourceType != "" {
 		d.Set(isSecurityGroupResourceType, *target.ResourceType)
 	}
+	if target != nil && *target.ResourceType == "load_balancer" {
+		_, waitErr := isWaitForSGTargetLBAvailable(sess, *target.ID, d.Timeout(schema.TimeoutCreate))
+		if waitErr != nil {
+			return fmt.Errorf("[ERROR] Error waiting for load balancer to become available while creating/updating Security Group Target Binding: %s", waitErr)
+		}
+	}
 
 	return nil
 }
@@ -222,6 +231,10 @@ func resourceIBMISSecurityGroupTargetDelete(d *schema.ResourceData, meta interfa
 		}
 		return fmt.Errorf("[ERROR] Error Getting Security Group Targets (%s): %s\n%s", securityGroupID, err, response)
 	}
+	// Acquire a lock based on the target ID to prevent simultaneous delete on same target
+	isSGTargetPrefixKey := "security_group_key_" + securityGroupTargetID
+	conns.IbmMutexKV.Lock(isSGTargetPrefixKey)
+	defer conns.IbmMutexKV.Unlock(isSGTargetPrefixKey)
 
 	deleteSecurityGroupTargetBindingOptions := sess.NewDeleteSecurityGroupTargetBindingOptions(securityGroupID, securityGroupTargetID)
 	response, err = sess.DeleteSecurityGroupTargetBinding(deleteSecurityGroupTargetBindingOptions)

--- a/ibm/service/vpc/resource_ibm_is_security_group_target.go
+++ b/ibm/service/vpc/resource_ibm_is_security_group_target.go
@@ -144,7 +144,7 @@ func resourceIBMISSecurityGroupTargetCreate(d *schema.ResourceData, meta interfa
 		if errsgt != nil {
 			return errsgt
 		}
-	} else if crn != nil && *crn != "" && strings.Contains(*crn, "virtual_network_interfaces") {
+	} else if crn != nil && *crn != "" && strings.Contains(*crn, "virtual-network-interface") {
 		vpcClient, err := meta.(conns.ClientSession).VpcV1API()
 		if err != nil {
 			return err

--- a/ibm/service/vpc/resource_ibm_is_security_group_target.go
+++ b/ibm/service/vpc/resource_ibm_is_security_group_target.go
@@ -196,12 +196,6 @@ func resourceIBMISSecurityGroupTargetRead(d *schema.ResourceData, meta interface
 	if target.ResourceType != nil && *target.ResourceType != "" {
 		d.Set(isSecurityGroupResourceType, *target.ResourceType)
 	}
-	if target != nil && *target.ResourceType == "load_balancer" {
-		_, waitErr := isWaitForSGTargetLBAvailable(sess, *target.ID, d.Timeout(schema.TimeoutCreate))
-		if waitErr != nil {
-			return fmt.Errorf("[ERROR] Error waiting for load balancer to become available while creating/updating Security Group Target Binding: %s", waitErr)
-		}
-	}
 
 	return nil
 }

--- a/ibm/service/vpc/resource_ibm_is_security_group_target_test.go
+++ b/ibm/service/vpc/resource_ibm_is_security_group_target_test.go
@@ -143,3 +143,99 @@ resource "ibm_is_security_group_target" "testacc_security_group_target" {
   }`, vpcname, subnetname, zoneName, cidr, name, lbname)
 
 }
+
+func TestAccIBMISSecurityGroupTarget_UpdatePendingRetry(t *testing.T) {
+	var securityGroup1, securityGroup2 string
+
+	vpcname := fmt.Sprintf("tfsg-vpc-%d", acctest.RandIntRange(10, 100))
+	subnetname := fmt.Sprintf("tfsg-subnet-%d", acctest.RandIntRange(10, 100))
+	lbname := fmt.Sprintf("tfsg-lb-%d", acctest.RandIntRange(10, 100))
+	sg1name := fmt.Sprintf("tfsg-one-%d", acctest.RandIntRange(10, 100))
+	sg2name := fmt.Sprintf("tfsg-two-%d", acctest.RandIntRange(10, 100))
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { acc.TestAccPreCheck(t) },
+		Providers:    acc.TestAccProviders,
+		CheckDestroy: testAccCheckIBMISSecurityGroupTargetDestroy,
+		Steps: []resource.TestStep{
+			// Create first security group target binding
+			{
+				Config: testAccCheckIBMISsecurityGroupTargetConfig(vpcname, subnetname, acc.ISZoneName, acc.ISCIDR, lbname, sg1name),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckIBMISSecurityGroupTargetExists("ibm_is_security_group_target.testacc_security_group_target", &securityGroup1),
+					resource.TestCheckResourceAttr(
+						"ibm_is_security_group_target.testacc_security_group_target", "name", lbname),
+					resource.TestCheckResourceAttrSet(
+						"ibm_is_security_group_target.testacc_security_group_target", "security_group"),
+					resource.TestCheckResourceAttrSet(
+						"ibm_is_security_group_target.testacc_security_group_target", "target"),
+					resource.TestCheckResourceAttrSet(
+						"ibm_is_security_group_target.testacc_security_group_target", "crn"),
+					resource.TestCheckResourceAttrSet(
+						"ibm_is_security_group_target.testacc_security_group_target", "resource_type"),
+				),
+			},
+			// Create second security group target binding for the same LB - should trigger UPDATE_PENDING condition
+			{
+				Config: testAccCheckIBMISsecurityGroupTargetDualConfig(vpcname, subnetname, acc.ISZoneName, acc.ISCIDR, lbname, sg1name, sg2name),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckIBMISSecurityGroupTargetExists("ibm_is_security_group_target.testacc_security_group_target", &securityGroup1),
+					testAccCheckIBMISSecurityGroupTargetExists("ibm_is_security_group_target.testacc_security_group_target2", &securityGroup2),
+					resource.TestCheckResourceAttr(
+						"ibm_is_security_group_target.testacc_security_group_target", "name", lbname),
+					resource.TestCheckResourceAttr(
+						"ibm_is_security_group_target.testacc_security_group_target2", "name", lbname),
+					resource.TestCheckResourceAttrSet(
+						"ibm_is_security_group_target.testacc_security_group_target2", "security_group"),
+					resource.TestCheckResourceAttrSet(
+						"ibm_is_security_group_target.testacc_security_group_target2", "target"),
+					resource.TestCheckResourceAttrSet(
+						"ibm_is_security_group_target.testacc_security_group_target2", "crn"),
+					resource.TestCheckResourceAttrSet(
+						"ibm_is_security_group_target.testacc_security_group_target2", "resource_type"),
+				),
+			},
+		},
+	})
+}
+
+// Config for dual security groups attached to the same load balancer
+func testAccCheckIBMISsecurityGroupTargetDualConfig(vpcname, subnetname, zoneName, cidr, lbname, sg1name, sg2name string) string {
+	return fmt.Sprintf(`
+		resource "ibm_is_vpc" "testacc_vpc" {
+			name = "%s"
+		}
+
+		resource "ibm_is_subnet" "testacc_subnet" {
+			name = "%s"
+			vpc = ibm_is_vpc.testacc_vpc.id
+			zone = "%s"
+			ipv4_cidr_block = "%s"
+		}
+
+		resource "ibm_is_security_group" "testacc_security_group_one" {
+			name = "%s"
+			vpc = ibm_is_vpc.testacc_vpc.id
+		}
+
+		resource "ibm_is_security_group" "testacc_security_group_two" {
+			name = "%s"
+			vpc = ibm_is_vpc.testacc_vpc.id
+		}
+
+		resource "ibm_is_lb" "testacc_LB" {
+			name = "%s"
+			subnets = [ibm_is_subnet.testacc_subnet.id]
+		}
+
+		resource "ibm_is_security_group_target" "testacc_security_group_target" {
+			security_group = ibm_is_security_group.testacc_security_group_one.id
+			target = ibm_is_lb.testacc_LB.id
+		}
+
+		resource "ibm_is_security_group_target" "testacc_security_group_target2" {
+			security_group = ibm_is_security_group.testacc_security_group_two.id
+			target = ibm_is_lb.testacc_LB.id
+		}`,
+		vpcname, subnetname, zoneName, cidr, sg1name, sg2name, lbname)
+}

